akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -511,13 +511,10 @@ private[stream] object Collect {
       }
 
       override def onUpstreamFinish(): Unit = {
-        if (current == zero) {
-          eventualCurrent.value match {
-            case Some(Success(`zero`)) ⇒
-              // #24036 upstream completed without emitting anything but after zero was emitted downstream
-              completeStage()
-            case _ ⇒ // in all other cases we will get a complete when the future completes
-          }
+        eventualCurrent.value match {
```

Comment:
`onRestart`, `current` is reset but `eventualCurrent` is not. Would that lead to problems when combined with this change?

Reply:
Good point. I was thinking about it, but this specific case was not solved in the previous implementation. I will try to come up with the test to simulate that situation.

Reply:
I think I constructed the test case which should not fail. Please review:

```
      val promises = List(Promise[Int](), Promise[Int]())
      promises(0).complete(Success(1))
      promises(1).complete(Failure(TE("bang")))

      val (pub, sub) =
        TestSource.probe[Int]
          .via(Flow[Int].scanAsync(0)((acc, in) ⇒ promises(in).future))
          .withAttributes(ActorAttributes.supervisionStrategy(Supervision.restartingDecider))
          .toMat(TestSink.probe)(Keep.both)
          .run()

      pub.sendNext(0)
      sub.request(10)
      sub.expectNext(0)
      sub.expectNext(1)
      pub.sendNext(1)
      sub.expectError(TE("bang"))
      pub.sendComplete()
      sub.expectComplete()
```

Am I right that it should be running fine?

Reply:
Please, disregard this test, I've pushed another fix and simplified the code, also covered some more supervision situations with tests.